rfcs/003 - Multi-Region Support RFC.md

```diff
@@ -0,0 +1,149 @@
+
+##### Pros
+
+1. Also requires zero schema changes to other modules, can be done in a single shot.
+2. Preserves the "default region" concept to keep inserts effectively identical.
+3. Can actually be faster by cutting out irrelevant regions.
+
+##### Cons
+
+1. Joins would be very slow and meta, with no good way to guarantee correctness as the joins will not be enforced by Postgres, but only us.
+2. Inserts for non-default regions are all two-step, and *must* be explicitly done before `iasql_apply` or it will accidentally create in the wrong region.
+3. Querying what's in what region will be much tougher than other options.
+
+#### Region-prefixed-to-@cloudId-column
+
+The IaSQL `@cloudId` concept (an ID that isn't the autogenerated database ID, but the ID the cloud provider generates once the resource is created) can double-up as the region indicator, when necessary. On insert you insert the region in question, and once it is created in the cloud, it it set to `region:cloudId` in the column.
+
+##### Pros
+
+1. This also requires no schema changes and can be updated over time.
+2. Could similarly have a `default_region` and `active_regions` concept like the mega-join table to gain its benefits.
+
+##### Cons
+
+1. Querying the records by region will be complicated with string prefix checking, so can never be performant, and also varies entity-by-entity.
+2. The cloud ID itself similarly would need to be parsed out to find the relevant part of the string, and so joins between entities *on* the cloudId become weirder.
+
+#### Optional region column on relevant tables
+
+Coming full circle on this, we can improve on the first option in this list by taking the `default_region` and `active_regions` idea from the prior two options and applying that to it.
+
+##### Pros
+
+1. Inserts *can* stay the same and not specify a region, with the mappers automatically assuming the default region in question.
+2. It can also be migrated to over time.
+
+##### Cons
+
+1. Nulling-out the region column after it has been populated can have unpredictable results, and is easier to accidentally do than the prior option.
+2. Most every entity needs to be altered.
+
+#### Default region column on relevant tables
+
+Minor tweak to the above, instead of the region being nullable, they are not-nullable but have a default value equal to the `default_region` column in the `aws_account`. This can be accomplished with a simple Postgres function that returns the `default_region` value.
+
+##### Pros
+
+1. Same insert benefits as the above, but also there's never weirdness with the region not being defined.
+
+##### Cons
+
+1. There may be an expectation that changing the default region changes the values of records with the same region?
```

Comment:
after reading all I'm leaning toward this option. We could also make the default region in aws_account as read-only, letting users just set it on db creation. They might want to change the default region but not all the resources to that region or as you said in the cons, a massive movement of resources from one region to another, but any of those would be an opinionated decision.

Reply:
It does feel like the best of them.

I think the default region should be mutable but not-NULLable and ideally dependent on a Regions table that we pull with their credentials, but all of those requirements produce a chicken-and-egg problem on how to store the creds before writing the default region.

If you're okay with this, though, I'll elevate this approach to the suggested one and out of the alternatives considered and flesh it out a bit more?

Reply:
definitely, something to think about. it might need to happen in a 3 step process behind scenes: 
- save cloud creds
- complete regions table
- save the default region


Yes I think we should go a bit deeper with this solution